From 9fb829591a9b0969893587a94269bfa8d336be56 Mon Sep 17 00:00:00 2001
From: David Cohen <david.a.cohen@intel.com>
Date: Thu, 22 May 2014 14:23:07 -0700
Subject: [PATCH 2/3] usb: dwc3: HACK: assert ulpi_refclk

In some BIOS ulpi refclk may be disabled, which causes usb phy to not
work.

This hackish patch makes kernel more robust against bozos from windows'
BIOS providing FW for poor Android guys.

Signed-off-by: David Cohen <david.a.cohen@intel.com>
---
 drivers/usb/dwc3/dwc3-pci.c | 37 +++++++++++++++++++++++++++++++++++++
 1 file changed, 37 insertions(+)

diff --git a/drivers/usb/dwc3/dwc3-pci.c b/drivers/usb/dwc3/dwc3-pci.c
index 740626f90e29..c481acc193c2 100644
--- a/drivers/usb/dwc3/dwc3-pci.c
+++ b/drivers/usb/dwc3/dwc3-pci.c
@@ -94,6 +94,36 @@ err1:
 	return ret;
 }
 
+#define GP_RWREG1			0xa0
+#define GP_RWREG1_ULPI_REFCLK_DISABLE	(1 << 17)
+static void dwc3_pci_enable_ulpi_refclock(struct pci_dev *pci)
+{
+	void __iomem	*reg;
+	struct resource	res;
+	struct device	*dev = &pci->dev;
+	u32 		value;
+
+	res.start	= pci_resource_start(pci, 1);
+	res.end 	= pci_resource_end(pci, 1);
+	res.name	= "dwc_usb3_bar1";
+	res.flags	= IORESOURCE_MEM;
+
+	reg = devm_ioremap_resource(dev, &res);
+	if (IS_ERR(reg)) {
+		dev_err(dev, "cannot check GP_RWREG1 to assert ulpi refclock\n");
+		return;
+	}
+
+	value = readl(reg + GP_RWREG1);
+	if (!(value & GP_RWREG1_ULPI_REFCLK_DISABLE))
+		return; /* ULPI refclk already enabled */
+
+	/* Let's clear ULPI refclk disable */
+	dev_warn(dev, "ULPI refclock is disable from the BIOS, let's try to enable it\n");
+	value &= ~GP_RWREG1_ULPI_REFCLK_DISABLE;
+	writel(value, reg + GP_RWREG1);
+}
+
 static int dwc3_pci_probe(struct pci_dev *pci,
 		const struct pci_device_id *id)
 {
@@ -158,6 +188,13 @@ static int dwc3_pci_probe(struct pci_dev *pci,
 	dwc3->dev.parent = dev;
 	glue->dwc3 = dwc3;
 
+	/*
+	 * HACK: we found an issue when enabling DWC3 controller where the
+	 * refclock to the phy is not being enabled.
+	 * We need an extra step to make sure such clock is enabled.
+	 */
+	dwc3_pci_enable_ulpi_refclock(pci);
+
 	ret = platform_device_add(dwc3);
 	if (ret) {
 		dev_err(dev, "failed to register dwc3 device\n");
-- 
1.9.1

